Sources/Fuzzilli/Core/CodeTemplate.swift

```diff
@@ -0,0 +1,100 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+public class CodeTemplate {
+    public let name: String
+
+    public var stats = ProgramGeneratorStats()
+
+    private let f: (ProgramBuilder) -> ()
+
+    init(_ name: String, _ f: @escaping (_: ProgramBuilder) -> ()) {
+        self.name = name
+        self.f = f
+    }
+
+    func run(in b: ProgramBuilder) {
+        f(b)
+    }
+
+    /// Generate an array of random function signatures
+    static func generateRandomFunctionSignatures(forFuzzer fuzzer: Fuzzer, n: Int) -> [FunctionSignature] {
+        var signatures = [FunctionSignature]()
+        signatures.append(CodeTemplate.generateSignature(forFuzzer: fuzzer, n: Int.random(in: 0...3)))
+        return signatures
+    }
+
+    /// This function generates and sets property types for the global properties
+    static func generateRandomPropertyTypes(forBuilder b: ProgramBuilder) {
+        for _ in 0..<5 {
```

Comment:
This should probably be a parameter `n` as well

Reply:
I think `5` is related to `customPropertyNames` if I am not mistaken? i.e. a, b, c, d, e?
We should probably just use `fuzzer.environment.customPropertyNames.count`? and then go through each of the properties and generate a random type instead of `chooseUniform`? WDYT?

Reply:
Oh hmm. I guess we'd want at least a few properties to have no types. How about we generate types for half of them?